_blog/2024-01-23-using-serde-rust.mdx

```diff
@@ -0,0 +1,341 @@
+    message: String,
+}
+
+fn to_and_from_json() {
+    let json = json!({"message": "Hello world!"});
+    let my_struct: MyStruct = serde_json::from_str(&json).unwrap();
+
+    assert_eq!(my_struct, MyStruct { message: "Hello world!".to_string());
+
+    assert!(serde_json::to_string(my_struct).is_ok());
+}
+```
+
+If you’re interested in using `serde-json` for your Rust application, we have an article talking about JSON parsing libraries which you can check out [here.](https://www.shuttle.rs/blog/2024/01/18/parsing-json-rust)  
+
+We can also deserialize and serialize to/from many sources including from a file stream I/O, a JSON byte array and more!
+
+### Implementing Deserialize and Serialize manually
+
+As a side note, you can also implement `Deserialize` and `Serialize` manually! This is quite complicated, but for now we will stick with a simple implementation. Here is a simple implementation for serializing of the `i32` primitive type. 
+
+```rust
+use serde::{Serializer, Serialize};
+
+impl Serialize for i32 {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        serializer.serialize_i32(*self)
+    }
+}
+```
+
+To be able to convert the type, `serde` internally requires us to use a type that implements `Serializer`.  To implement `Serialize` for a type that isn’t directly a primitive, we can extend this by serializing into a primitive, then converting into whatever type we want from the primitive. If we want custom serialization for structs, we can also do the same with use of the `SerializeStruct` trait:
+
+```rust
+use serde::ser::{Serialize, Serializer, SerializeStruct};
+
+struct Color {
+    r: u8,
+    g: u8,
+    b: u8,
+}
+
+impl Serialize for Color {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        // 3 is the number of fields in the struct.
+        let mut state = serializer.serialize_struct("Color", 3)?;
+        state.serialize_field("r", &self.r)?;
+        state.serialize_field("g", &self.g)?;
+        state.serialize_field("b", &self.b)?;
+        state.end()
+    }
+}
+```
+Note that to serialize a field, the field type also needs to implement `Serialize`. If you have a custom type that doesn't implement `Serialize`, you will either need to implement `Serialize` or use a `Serialize` derive macro (if the struct/enum type holds types that all implement `Serialize`).
+
+The `Deserialize` trait is a little bit different and is a fair bit more complicated to implement. To be able to deserialize to a type, the type itself needs to implement `Sized` which means that there are a number of types which can’t use this trait (for example `&str`) because they are unsized types. To deserialize a type, you also need to use a type that implements the `Visitor` trait. The `serde` documentation provides the below example for an `I32Visitor` type that attempts to deserialize multiple types to `i32`:
```

Comment:
Do we have an article on `Sized` ? This might be an interesting topic to cover. Maybe with `Send` and `Sync` as well, if simply covering `Sized` isn't enough.

Reply:
Not yet but you do actually bring up a pretty good point